extensions/picocli/runtime/src/main/java/io/quarkus/picocli/runtime/PicocliCommandLineProducer.java

```diff
@@ -0,0 +1,63 @@
+package io.quarkus.picocli.runtime;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.literal.NamedLiteral;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import io.quarkus.arc.DefaultBean;
+import io.quarkus.picocli.runtime.annotations.TopCommand;
+import picocli.CommandLine;
+
+@ApplicationScoped
+public class PicocliCommandLineProducer {
+
+    private final Instance<Object> topCommand;
+    private final PicocliConfiguration picocliConfiguration;
+
+    @Inject
+    public PicocliCommandLineProducer(@TopCommand Instance<Object> topCommand, PicocliConfiguration picocliConfiguration) {
+        this.topCommand = topCommand;
+        this.picocliConfiguration = picocliConfiguration;
+    }
+
+    private Class<?> classForName(String name) {
+        try {
+            return Class.forName(name);
+        } catch (ClassNotFoundException ex) {
+            throw new IllegalArgumentException(ex);
+        }
+    }
+
+    @Produces
+    @DefaultBean
+    @Singleton
+    public CommandLine.IFactory picocliFactory() {
+        return new PicocliBeansFactory();
+    }
+
+    @Produces
+    @DefaultBean
+    public PicocliCommandLineFactory picocliCommandLineFactory(CommandLine.IFactory picocliFactory) {
+        return () -> picocliConfiguration.topCommand
```

Comment:
We generally try to avoid using lamdbas in runtime code as much as possible, due to their non-zero (albeit very small) runtime overhead.